src/grpc/qgrpcchannel.cpp

```diff
@@ -0,0 +1,382 @@
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/impl/codegen/byte_buffer.h>
+#include <grpcpp/impl/codegen/channel_interface.h>
+#include <grpcpp/impl/codegen/client_callback_impl.h>
+#include <grpcpp/impl/codegen/client_context.h>
+#include <grpcpp/impl/codegen/client_unary_call.h>
+#include <grpcpp/impl/codegen/completion_queue.h>
+#include <grpcpp/impl/codegen/rpc_method.h>
+#include <grpcpp/impl/codegen/slice.h>
+#include <grpcpp/impl/codegen/status.h>
+#include <grpcpp/impl/codegen/status_code_enum.h>
+#include <grpcpp/impl/codegen/sync_stream_impl.h>
+#include <grpcpp/security/credentials.h>
+
+#include "qabstractgrpccredentials.h"
+#include "qgrpcasyncreply.h"
+#include "qgrpcsubscription.h"
+#include "qabstractgrpcclient.h"
+#include "qgrpccredentials.h"
+#include "qprotobufserializerregistry_p.h"
+#include "qtprotobuflogging.h"
+
+using namespace QtProtobuf;
+
+namespace QtProtobuf {
+
+class QGrpcChannelWorker;
+
+static grpc::Status parseByteBuffer(const grpc::ByteBuffer &buffer, QByteArray *data) {
+    std::vector<grpc::Slice> slices;
+    auto status = buffer.Dump(&slices);
+
+    for (auto slice : slices) {
+        const uint8_t *i = slice.begin();
+
+        while (i < slice.end()) {
+            data->append(*i);
+            i++;
+        }
+    }
+
+    return status;
+}
+
+//! \private
+class QGrpcChannelCall : public QObject {
+    //! \private
+    Q_OBJECT;
+public:
+    QGrpcChannelCall(QObject *parent = nullptr) : QObject(parent) {}
+
+    QByteArray response;
+    QGrpcStatus status;
+
+signals:
+    void finished();
+};
+
+class QGrpcChannelSubscription : public QObject {
+    Q_OBJECT;
+
+    friend class QGrpcChannelWorker;
+
+public:
+    QGrpcChannelSubscription(QObject *parent = nullptr) : QObject(parent) {
+        thread = QThread::create([this](){
+            grpc::ByteBuffer response;
+            QByteArray data;
+            grpc::Status status;
+
+            while (reader->Read(&response)) {
+                status = parseByteBuffer(response, &data);
+
+                if (!status.ok())
+                    goto end;
+
+                emit this->data(data);
+            }
+
+            status = reader->Finish();
+
+end:
+            this->status = {
+                (QGrpcStatus::StatusCode) status.error_code(),
+                status.error_message().c_str()
+            };
+
+            emit finished();
+        });
+    }
+
+    ~QGrpcChannelSubscription() {
+        terminate();
+    }
+
+	void start() {
+        thread->start();
+	}
+
+	void terminate() {
+        // TODO: check thread safety
+        qProtoDebug() << "Subscription thread terminated";
+        context.TryCancel();
+    }
+
+signals:
+    void data(const QByteArray &data);
+    void finished();
+
+public:
+    QGrpcStatus status;
+
+private:
+    QThread *thread;
+    grpc::ClientContext context;
+    grpc_impl::ClientReader<grpc::ByteBuffer> *reader;
+};
+
+//! \private
+class QGrpcChannelWorker : public QObject {
```

Comment:
Could you tell me briefly how you would use QFuture or std::future for the worker class, since I have no idea?

FYI my implementation comes from Qt documentation, which suggests creating a Worker class and calling moveToThread()

Reply:
Future it self is already implementer worker thread. So idea is to create future and wait until opration is finished, either in same stack or subscribe to it finish. Overall we may keep current implementation for now